File: 1381. Design a Stack With Increment Operation/solution.js
/**
 * @param {number} maxSize
 */
var CustomStack = function(maxSize) {
    this.nums = [];
    this.mxSize = maxSize;
};

/** 
 * @param {number} x
 * @return {void}
 */
CustomStack.prototype.push = function(x) {
     if (this.nums.length === this.mxSize) return;
        this.nums.push(x);
};

/**
 * @return {number}
 */
CustomStack.prototype.pop = function() {
     if (this.nums.length === 0) return -1;
        return this.nums.pop();
};

/** 
 * @param {number} k 
 * @param {number} val
 * @return {void}
 */
CustomStack.prototype.increment = function(k, val) {
     for (let i = 0; i < k && i < this.nums.length; i++) {
            this.nums[i] += val;
      }  
};

File: 1381. Design a Stack With Increment Operation/solution.c
typedef struct {
    int *nums;
    int mxSize;
    int topIndex;
} CustomStack;


CustomStack* customStackCreate(int maxSize) {
    CustomStack* obj = (CustomStack*)malloc(sizeof(CustomStack));
    obj->mxSize = maxSize;
    obj->nums = (int*)malloc(maxSize * sizeof(int));
    obj->topIndex = -1;
    return obj;
}

void customStackPush(CustomStack* obj, int x) {
    if (obj->topIndex == obj->mxSize - 1)
        return;
    obj->nums[++(obj->topIndex)] = x;
}

int customStackPop(CustomStack* obj) {
     if (obj->topIndex == -1)
        return -1;
    return obj->nums[(obj->topIndex)--];
}

void customStackIncrement(CustomStack* obj, int k, int val) {
    for (int i = 0; i < k && i <= obj->topIndex; i++) {
        obj->nums[i] += val;
    }
}

void customStackFree(CustomStack* obj) {
    free(obj->nums);
    free(obj);
}

